osgar/drivers/rosmsg.py

```diff
@@ -0,0 +1,228 @@
+    intensities = struct.unpack_from('<' + 'f' * intensities_size, data, pos)
+
+    # conversion to int and millimeters
+    scan = [int(x*1000) if x < 65.0 else 0 for x in scan]
+    return scan
+
+
+def parse_odom(data):
+    # http://docs.ros.org/melodic/api/nav_msgs/html/msg/Odometry.html
+    size = struct.unpack_from('<I', data)[0]
+    assert size == 719, size  # expected size for short lidar
+    pos = 4
+    seq, timestamp_sec, timestamp_nsec, frame_id_size = struct.unpack_from('<IIII', data, pos)
+    pos += 4 + 4 + 4 + 4
+    frame_id = data[pos:pos+frame_id_size]
+#    print(frame_id, timestamp_sec, timestamp_nsec)
+    pos += frame_id_size
+    child_frame_id = struct.unpack_from('<I', data, pos)[0]
+    pos += 4
+    #print(data[pos:pos + child_frame_id])
+    pos += child_frame_id
+    # http://docs.ros.org/melodic/api/geometry_msgs/html/msg/PoseWithCovariance.html
+    # http://docs.ros.org/melodic/api/geometry_msgs/html/msg/Pose.html
+    # http://docs.ros.org/melodic/api/geometry_msgs/html/msg/Point.html
+    # http://docs.ros.org/melodic/api/geometry_msgs/html/msg/Quaternion.html
+    x, y, z = struct.unpack_from('<ddd', data, pos)
+    pos += 3 * 8
+    ori = struct.unpack_from('<dddd', data, pos)
+    pos += 4 * 8
+    # float64[36] covariance
+    pos += 36 * 8
+    #print(x, y, z)
+
+    # http://docs.ros.org/melodic/api/geometry_msgs/html/msg/Twist.html
+    linear = struct.unpack_from('<ddd', data, pos)
+    pos += 3 * 8
+    angular = struct.unpack_from('<ddd', data, pos)
+    pos += 3 * 8
+    # float64[36] covariance
+    pos += 36 * 8
+    assert pos == len(data), (pos, len(data))
+
+    q0, q1, q2, q3 = ori  # quaternion
+    ax =  math.atan2(2*(q0*q1+q2*q3), 1-2*(q1*q1+q2*q2))
+    ay =  math.asin(2*(q0*q2-q3*q1))
+    az =  math.atan2(2*(q0*q3+q1*q2), 1-2*(q2*q2+q3*q3))
+    return (x, y, ax)
+
+
+class ROSMsgParser(Thread):
+    def __init__(self, config, bus):
+        Thread.__init__(self)
+        self.setDaemon(True)
+
+        self.bus = bus
+        self._buf = b''
+
+        self.topic_type = config.get('topic_type')
+
+    def get_packet(self):
+        data = self._buf
+        if len(data) < 4:
+            return None
+        size = struct.unpack_from('<I', data, 0)[0]
+#        print(size, len(self._buf))
+        assert size > 0, size
+        size += 4  # the length prefix
+        if len(data) < size:
+            return None
+        ret, self._buf = data[:size], data[size:]        
+        return ret
+
+    def run(self):
+        try:
+            header = None
+            # initial message contains structure
+            while True:
+                timestamp, channel, data = self.bus.listen()
+                self._buf += data
+                packet = self.get_packet()
+                if packet is not None:
+                    if header is None:
+                        header = packet
+                    elif self.topic_type == 'sensor_msgs/CompressedImage':
```

Comment:
Hmm. So. One instance of this node supports parsing one topic type. And based on the type of the topic, it selects the name of the output where it publishes. That seems strange.